### PR TITLE
Fix ibd shared past negotiation to be non quadratic also in the worst-case

### PR DESCRIPTION
**Pull request by michaelsutton**

### CODE REVIEW

File: app/protocol/flows/v5/blockrelay/ibd.go [hunk 1]
```diff
@@ -338,10 +348,17 @@ func (flow *handleIBDFlow) syncPruningPointFutureHeaders(consensus externalapi.C
 	}
 }
 
-func (flow *handleIBDFlow) sendRequestHeaders(highestSharedBlockHash *externalapi.DomainHash,
-	peerSelectedTipHash *externalapi.DomainHash) error {
+func (flow *handleIBDFlow) sendRequestPastDiff(
+	syncerHeaderSelectedTipHash, relayBlockHash *externalapi.DomainHash) error {
+
+	msgGetPastDiff := appmessage.NewMsgRequestPastDiff(syncerHeaderSelectedTipHash, relayBlockHash)
```
Comment (someone235) [inline app/protocol/flows/v5/blockrelay/ibd.go:354]: Rename `Get` to `Request`

File: app/protocol/flows/v5/blockrelay/ibd.go [hunk 2]
```diff
@@ -338,10 +348,17 @@ func (flow *handleIBDFlow) syncPruningPointFutureHeaders(consensus externalapi.C
...
+	return flow.outgoingRoute.Enqueue(msgGetPastDiff)
+}
+
+func (flow *handleIBDFlow) sendRequestHeaders(
+	highestKnownSyncerChainHash, syncerHeaderSelectedTipHash *externalapi.DomainHash) error {
 
-	msgGetBlockInvs := appmessage.NewMsgRequstHeaders(highestSharedBlockHash, peerSelectedTipHash)
+	msgGetBlockInvs := appmessage.NewMsgRequstHeaders(highestKnownSyncerChainHash, syncerHeaderSelectedTipHash)
```
Comment (someone235) [inline app/protocol/flows/v5/blockrelay/ibd.go:361]: Rename Get to Request

File: app/protocol/flows/v5/blockrelay/ibd_with_headers_proof.go [hunk 3]
```diff
@@ -45,10 +46,26 @@ func (flow *handleIBDFlow) ibdWithHeadersProof(highHash *externalapi.DomainHash,
...
-	if !highestSharedBlockFound {
+	var highestSharedBlockFound, isPruningPointInSharedBlockChain bool
+	if highestSharedBlockHash != nil {
+		highestSharedBlockFound = true
+		pruningPoint, err := flow.Domain().Consensus().PruningPoint()
+		if err != nil {
+			return false, false, err
+		}
+
+		isPruningPointInSharedBlockChain, err = flow.Domain().Consensus().IsInSelectedParentChainOf(
```
Comment (someone235) [inline app/protocol/flows/v5/blockrelay/ibd_with_headers_proof.go:61]: If this is false, doesn't it mean we have a finality conflict?

File: domain/consensus/processes/syncmanager/syncmanager.go [hunk 4]
```diff
@@ -69,6 +70,32 @@ func (sm *syncManager) GetHashesBetween(stagingArea *model.StagingArea, lowHash,
...
+	isRequestedAncestorOfHas, err := sm.dagTopologyManager.IsAncestorOf(stagingArea, requestedHash, hasHash)
+	if err != nil {
+		return nil, err
+	}
+	if isRequestedAncestorOfHas {
+		return nil, errors.Errorf("expected block %s to be in anticone of %s",
+			hasHash,
+			requestedHash)
+	}
+	// TODO: use maxBlocks limit
```
Comment (someone235) [inline domain/consensus/processes/syncmanager/syncmanager.go:95]: This must be handled before merge

**Review by someone235 — CHANGES_REQUESTED** [review]

File: app/appmessage/p2p_msgrequestpastdiff.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+import (
+	"github.com/kaspanet/kaspad/domain/consensus/model/externalapi"
+)
+
+// MsgRequestPastDiff implements the Message interface and represents a kaspa
+// RequestHeaders message. It is used to request a past diff past(RequestedHash) \setminus past(HasHash)
+type MsgRequestPastDiff struct {
+	baseMessage
+	HasHash       *externalapi.DomainHash
```
[the remark this answers is not in the document]
Comment (michaelsutton) [reply app/appmessage/p2p_msgrequestpastdiff.go:15]: It's the hash the syncee already has including its past. Any other name ideas? ExistingHash?

File: app/protocol/flows/v5/blockrelay/ibd_with_headers_proof.go (hunk 3 above)
Comment (michaelsutton) [reply app/protocol/flows/v5/blockrelay/ibd_with_headers_proof.go:61]: This is exactly the same case where with previous algorithm we got shared (chain) block not found

File: domain/consensus/processes/syncmanager/syncmanager.go (hunk 4 above)
Comment (michaelsutton) [reply domain/consensus/processes/syncmanager/syncmanager.go:95]: How do you suggest doing it? Should I propegate a max lim into AnticoneFromBlock?

Comment (michaelsutton) [reply domain/consensus/processes/syncmanager/syncmanager.go:95]: I mean, I don't want to enter a long traversal if the syncee is dishonest

File: app/protocol/flows/v5/blockrelay/handle_request_past_diff.go [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+		hasHash, requestedHash, err := receiveRequestPastDiff(flow.incomingRoute)
+		if err != nil {
+			return err
+		}
+		log.Debugf("Received requestPastDiff with hasHash: %s, requestedHash: %s", hasHash, requestedHash)
+		log.Debugf("Getting past(%s) setminus past(%s) to %s", requestedHash, hasHash, flow.peer)
+
+		// GetPastDiff is a relatively heavy operation so we limit it
+		// in order to avoid locking the consensus for too long
+		// maxBlocks MUST be >= MergeSetSizeLimit + 1
```
[the remark this answers is not in the document]
Comment (michaelsutton) [reply app/protocol/flows/v5/blockrelay/handle_request_past_diff.go:48]: I'll rewrite the comment when I apply the limit

File: app/appmessage/p2p_msgrequestpastdiff.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (someone235) [reply app/appmessage/p2p_msgrequestpastdiff.go:15]: Maybe just neutral name? Maybe `blockHash` for `HasHash` and `context` for the other one, and call the message `RequestAnticone`?

[the remark this answers is not in the document]
Comment (michaelsutton) [reply app/appmessage/p2p_msgrequestpastdiff.go:15]: ok cool, makes sense

File: app/protocol/flows/v5/blockrelay/handle_request_anticone.go [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+			blockHeaders[i] = appmessage.DomainBlockHeaderToBlockHeader(blockHeader)
+		}
+
+		// We sort the headers in bottom-up topological order before sending
+		sort.Slice(blockHeaders, func(i, j int) bool {
+			return blockHeaders[i].BlueWork.Cmp(blockHeaders[j].BlueWork) < 0
+		})
+		if err != nil {
+			return err
+		}
```
Comment (someone235) [inline app/protocol/flows/v5/blockrelay/handle_request_anticone.go:73]: This looks redundant

File: domain/consensus/processes/dagtraversalmanager/anticone.go [hunk 8]
```diff
@@ -48,6 +50,13 @@ func (dtm *dagTraversalManager) AnticoneFromBlocks(stagingArea *model.StagingAre
 			return nil, err
 		}
 
+		// We count the number of blocks in past(tips) \setminus past(blockHash).
+		// We don't use `len(visited)` since it includes some maximal blocks in past(blockHash) as well.
+		traversalCounter++
+		if maxTraversalAllowed > 0 && traversalCounter > maxTraversalAllowed {
+			return nil, model.ErrReachedMaxTraversalAllowed
```
Comment (someone235) [inline domain/consensus/processes/dagtraversalmanager/anticone.go:57]: The best practice is to wrap errors with `errors.Wrapf`. When you do that, it adds stack information to the error

**Review by someone235 — CHANGES_REQUESTED** [review]

**Review by someone235 — APPROVED** [review]

**Pull request merged**